worker/peergrouper/worker.go

```diff
@@ -608,19 +479,66 @@ func (w *pgWorker) updateReplicaSet() (map[string]replicaset.Member, error) {
 		logger.Infof("successfully updated replica set")
 	}
 	if err := setHasVote(removed, false); err != nil {
-		return nil, errors.Annotate(err, "cannot set HasVote removed")
+		return nil, errors.Annotate(err, "removing non-voters")
+	}
+
+	// Return the members from the calculated peer-group that were in the
+	// original known replica-set members.
+	for i := range members {
+		rec := false
+		for j := range info.recognised {
+			if i == j {
+				rec = true
+				break
+			}
+		}
+		if !rec {
+			delete(members, i)
+		}
+	}
+	return members, nil
+}
+
+func prettyReplicaSetMembers(members map[string]*replicaset.Member) string {
+	var result []string
+	for _, m := range members {
+		result = append(result, fmt.Sprintf("    Id: %d, Tags: %v, Vote: %v", m.Id, m.Tags, isVotingMember(m)))
+	}
+	return strings.Join(result, "\n")
+}
+
+// peerGroupInfo collates current session information about the
+// mongo peer group with information from state machines.
+func (w *pgWorker) peerGroupInfo() (*peerGroupInfo, error) {
+	sts, err := w.config.MongoSession.CurrentStatus()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot get replica set status")
 	}
 
-	mm, _, _ := info.membersMap()
-	machineMembers := make(map[string]replicaset.Member)
-	for m, member := range mm {
-		machineMembers[m.Id()] = *member
+	mbr, err := w.config.MongoSession.CurrentMembers()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot get replica set members")
 	}
-	return machineMembers, nil
+
+	sp, err := w.getHASpaceFromConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return newPeerGroupInfo(w.machineTrackers, sts.Members, mbr, w.config.MongoPort, sp)
 }
 
-// setHasVote sets the HasVote status of all the given
-// machines to hasVote.
+// getHASpaceFromConfig returns a SpaceName from the controller config for
+// HA space. If unset, the empty space ("") will be returned.
+func (w *pgWorker) getHASpaceFromConfig() (network.SpaceName, error) {
```

Comment:
this feels a bit like it should be a State.HASpace() function instead of a method of pgWorker.